bin/propolis-server/src/lib/serial/backend.rs

```diff
@@ -0,0 +1,397 @@
+    }
+}
+
+impl ClientHandle {
+    /// Attempts to write the bytes in `buf` to the backend.
+    ///
+    /// The backend may buffer some of the written bytes before sending them to
+    /// its device, so when this function returns, it is not guaranteed that all
+    /// of the written bytes have actually reached the guest.
+    ///
+    /// # Return value
+    ///
+    /// - `Ok(bytes)` if the write succeeded; `bytes` is the number of bytes
+    ///   that were written.
+    /// - `Err(ErrorKind::PermissionDenied)` if this handle does not grant write
+    ///   access to the device.
+    /// - `Err(ErrorKind::ConnectionAborted)` if this handle's client was
+    ///   previously disconnected from the backend, e.g. because it had a full
+    ///   read channel and was using the close-on-full-channel read discipline.
+    ///
+    /// # Cancel safety
+    ///
+    /// The future returned by this function is cancel-safe. If it is dropped,
+    /// it is guaranteed that no bytes were written to the backend or its
+    /// device. See [`SinkBuffer::write`].
+    pub(super) async fn write(
+        &mut self,
+        buf: &[u8],
+    ) -> Result<usize, std::io::Error> {
+        if buf.is_empty() {
+            return Ok(0);
+        }
+
+        if !matches!(self.permissions, Permissions::ReadWrite) {
+            return Err(std::io::Error::from(
+                std::io::ErrorKind::PermissionDenied,
+            ));
+        }
+
+        // The client handle may have been invalidated if it used the "close on
+        // full channel" discipline. If this client is no longer active, return
+        // an error.
+        //
+        // The read dispatcher task doesn't synchronize with writes, so it's
+        // possible for the client to be invalidated after the lock is dropped.
+        // (It can't be held while writing to the backend because
+        // `SinkBuffer::write` is async.) This is OK; since reads and writes
+        // aren't synchronized to begin with, in any situation like this it was
+        // equally possible for the write to finish before the client was
+        // invalidated.
+        {
+            let inner = self.backend.inner.lock().unwrap();
+            if !inner.clients.contains_key(&self.id) {
+                return Err(std::io::Error::from(
+                    std::io::ErrorKind::ConnectionAborted,
+                ));
+            }
+        }
+
+        Ok(self
+            .backend
+            .sink_buffer
+            .write(buf, self.backend.sink.as_ref())
+            .await
+            .expect("chardev sink writes always succeed"))
+    }
+}
+
+/// Character backend state that's protected by a lock.
+struct Inner {
+    /// A history of the bytes read from this backend's device.
+    buffer: HistoryBuffer,
+
+    /// A table mapping client IDs to clients.
+    clients: BTreeMap<ClientId, Client>,
+
+    /// The ID to assign to the next client to attach to this backend.
+    next_client_id: u64,
```

Comment:
Rather than a `BTreeMap` and counter of IDs, I might consider using the [`slab`](https://crates.io/crates/slab) crate, which implements a collection of "things" indexed by integer ID, but stored in an array (so both iteration and lookups are likely much more efficient). Do note that `slab` will reuse indices of removed elements, in case that's an issue here, but AFAICT, it wouldn't actually be.